src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfoSourceCallback.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.TestCommon.Runners;
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Tools.Monitor;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit.Abstractions;
+
+namespace Microsoft.Diagnostics.Monitoring.Tool.UnitTests
+{
+    internal sealed class EndpointInfoSourceCallback : IEndpointInfoSourceCallbacks
+    {
+        private readonly ITestOutputHelper _outputHelper;
+        /// <summary>
+        /// Used to protect the completion list from mutation while processing
+        /// callbacks from it. The processing is done in an async method with async
+        /// calls, which are not allowed in a lock, thus use SemaphoreSlim.
+        /// </summary>
+        private readonly SemaphoreSlim _completionEntriesSemaphore = new(1);
+        private readonly List<CompletionEntry> _completionEntries = new();
+
+        public EndpointInfoSourceCallback(ITestOutputHelper outputHelper)
+        {
+            _outputHelper = outputHelper;
+        }
+
+        public async Task<IEndpointInfo> WaitForNewEndpointInfoAsync(AppRunner runner, TimeSpan timeout)
```

Comment:
I don't believe we're using this param (`timeout`). Should we be for the `timeoutCancellation`?

Reply:
Good catch; I had it correct in my other branch but the merge got slightly messed up.